File: apps/desktop/src/lib/file-explorer/network/reconnect-error.ts
/** Carrying a typed `ReconnectError` from the SMB reconnect commands to the log that records it. */
import type { ReconnectError } from '$lib/ipc/bindings'
import { TypedFailure, failureOf } from '$lib/ipc/typed-failure'

/**
 * An `Error` that still carries the backend's typed reconnect refusal.
 *
 * Its `Error.message` is a compact diagnostic (`reconnect refused: volume`), so
 * the reconnect manager's existing log lines stay readable without anyone
 * reading a backend sentence.
 */
export class ReconnectFailure extends TypedFailure<ReconnectError> {
  constructor(failure: ReconnectError) {
    super(failure, `reconnect refused: ${failure.type}`)
    this.name = 'ReconnectFailure'
  }
}

/** Throws a wire `ReconnectError` as an `Error`, keeping the typed value. */
export function throwReconnectError(failure: ReconnectError): never {
  throw new ReconnectFailure(failure)
}

/** The typed refusal behind a caught value, or `null` when it isn't one. */
export function asReconnectError(error: unknown): ReconnectError | null {
  return failureOf(ReconnectFailure, error)
}

/**
 * A one-line diagnostic for the log: the refusal's REASON, never a sentence.
 *
 * ❌ Not user-facing. The reconnect surfaces (the reconnecting view, the gave-up
 * banner, the sign-in form) carry their own translated copy; this exists so a
 * log line records what happened without anyone parsing prose to find out.
 */
export function describeReconnectRefusal(error: ReconnectError): string {
  return error.type === 'volumeNotFound' ? `volumeNotFound(${error.volumeId})` : `volume/${error.error.type}`
}

File: apps/desktop/src-tauri/src/file_system/volume/reconnect_error.rs
//! Why rebuilding a network share's session didn't work, as a value rather than
//! a sentence.
//!
//! ❌ Nothing here is prose a user reads. The reconnect surfaces speak for
//! themselves (the reconnecting view, the gave-up banner, the sign-in form's
//! own copy); this type is what the frontend LOGS and what a future surface
//! would render from. The volume's own refusal rides through
//! [`ReconnectError::Volume`] carrying the whole `VolumeError`, which is
//! already the wire type, so a backend that grows a variant reaches the
//! frontend without a second vocabulary to keep in step.

use cmdr_fs::volume::VolumeError;
use serde::{Deserialize, Serialize};

/// A typed refusal from `reconnect_smb_volume` /
/// `reconnect_smb_volume_with_credentials`.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
#[serde(tag = "type", rename_all = "camelCase", rename_all_fields = "camelCase")]
pub enum ReconnectError {
    /// The id isn't registered in `VolumeManager` any more (a race: the share
    /// was unmounted between the frontend's tick and the backend acting on it).
    VolumeNotFound {
        /// The id that no longer resolves.
        volume_id: String,
    },
    /// The volume refused, and said why in its own vocabulary. A non-SMB volume
    /// answers `VolumeError::NotSupported` here (the trait default); the
    /// frontend only ever calls this for known SMB volumes.
    Volume {
        /// The backend's typed answer.
        error: VolumeError,
    },
}

impl std::fmt::Display for ReconnectError {
    /// ❗ For logs and debugging only.
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::VolumeNotFound { volume_id } => write!(f, "volume not found: {volume_id}"),
            Self::Volume { error } => write!(f, "volume: {error}"),
        }
    }
}

impl std::error::Error for ReconnectError {}

impl From<VolumeError> for ReconnectError {
    fn from(error: VolumeError) -> Self {
        Self::Volume { error }
    }
}
